### ISSUE TITLE
Limited number of Nextcloud apps available

### ISSUE THREAD

**Issue by provokateurin** [NixOS/nixpkgs#332780]
## Issue description

The available Nextcloud apps are restricted by https://github.com/NixOS/nixpkgs/blob/master/pkgs/servers/nextcloud/packages/nextcloud-apps.json and https://github.com/NixOS/nixpkgs/blob/eec16b48aa9e98c64b4fdd4a564092ea03a9c8a8/pkgs/servers/nextcloud/packages/generate.sh#L11, but I fail to understand the reason why.

Only making a few apps available makes it a lot harder to install apps outside of those.

If it is for the licenses, then those could be automatically translated from the licenses available in https://github.com/helsinki-systems/nc4nix/blob/main/29.json (I'd be willing to help with that).

CC @dotlambda @Ma27 @pyrox0

**Comment by dotlambda** [comment]
There's also the issue that some apps need to be patched to work on NixOS, e.g. https://github.com/NixOS/nixpkgs/pull/211631.

**Comment by provokateurin** [comment]
I see, but that probably concerns the fewest of apps, no? Maybe then the better way is to remove apps that need to be patched (so basically invert the list).

**Comment by Ma27** [comment]
There's another issue with Nextcloud apps: you have to manually find out which license they use. Yes, there's a metadata field in the `appinfo.xml`, but it's worthless for us, because most app authors fail to differentiate between *Only and *Later (that's a common theme for the GPL family of licenses).

Also, I don't consider this too critical since there are at least three ways of installing apps into Nextcloud:
* use `fetchNextcloudAPp` yourself
* do it imperatively via the appstore (or whatever else it's called)
* I think the upstream nc4nix automatically generates all of the apps

Finally, nobody has stepped up to do that (and maintain the stuff!).

**Comment by provokateurin** [comment]
I get your point, but I still think it would be more convenient to have more apps available this way.

As for the license only vs. later the problem is that the allowed values are limited in https://github.com/nextcloud/appstore/blob/0681b16dc3fc6ac4bcce53a6c6ec1084a9f6c1d0/nextcloudappstore/api/v1/release/info.xsd#L368 and there is no way to differentiate them. It's not really the individual app developers that are failing to provide this information.
The reason for this is unclear to me, it seems the code was written a long time ago by an now ex-employee. This was probably an oversight the author didn't think about.

Another problem I see: It is possible to have an app with multiple licenses (not sure if that means dual-license or just parts with different licenses) and I don't think it is possible to provide multiple licenses in nix?

**Comment by SuperSandro2000** [comment]
> I get your point, but I still think it would be more convenient to have more apps available this way.

New apps can always be added via a PR. We are not limiting ourselves to those.

> The reason for this is unclear to me, it seems the code was written a long time ago by an now ex-employee. This was probably an oversight the author didn't think about.

Sure but getting that into shape for all apps is probably a lengthy effort.

> Another problem I see: It is possible to have an app with multiple licenses (not sure if that means dual-license or just parts with different licenses) and I don't think it is possible to provide multiple licenses in nix?

#335595

**Comment by provokateurin** [comment]
Regarding automated license mapping there is also https://github.com/nextcloud/appstore/issues/1274 and it seems like everyone is fine with transitioning to SPDX identifiers. It will just take the app authors some time to do that (and unmaintained apps are not going to get updated :/ )

**Comment by provokateurin** [comment]
Btw https://github.com/nextcloud/appstore/issues/1274 got implemented, so apps can now use SPDX identifiers in their appinfo/info.xml.
Should we re-open this issue to take another look at how we can now automatically parse this info to make apps available without manually adding them to the list?
From previous comments it seems like we still need to have some kind of blocklist for apps that don't work OOTB like recognize, but that should be doable.

**Comment by Ma27** [comment]
So, I had an interesting chat a few days ago with @britter where I got asked how we test nextcloud apps: what I basically do is checking if they all build and if we can package those w/o additional customizations (such as recognize) and that's pretty much it since it's pretty safe to assume that these will just work with our integration.
So the guarantee we effectively give is, it's as functional (or as broken) as in a stock Nextcloud installation.

Now, one may argue that distributions should provide a somewhat curated package-set which is something I don't consider realistic here at all. If we're all fine with that (I am!), I'd say carry on. Mostly bringing this up because I talked about it recently and to see if others disagree (cc @dotlambda @bachp @onny). 

> and unmaintained apps are not going to get updated

I assume these get rotated out of our package set when they don't update `appinfo/info.xml` anymore (because of being unmaintained), right? Is there another way to derive that an app is unmaintained from the metadata? Because if an app is actually unmaintained, that would still be a reason for me to drop it.

> apps that don't work OOTB like recognize, but that should be doable.

Is there any way to find out if something has external dependencies (even if these get magically installed by default) in some of the appstore metadata? Didn't find anything helpful in `info.xml`.
Also, what's the growth rate of Nextcloud the appstore btw? If we're talking about <3 new apps per update from us we'd have to check for external dependencies, I might be OK with that. But for more I'm not sure if we'll actually want to do that (happy to get more opinions, especially on that).

**Comment by provokateurin** [comment]
> I assume these get rotated out of our package set when they don't update appinfo/info.xml anymore

If they are not maintained then they are not going to update the compatible server versions and therefore will disappear at some point when the latest release doesn't even match the lowest supported server version anymore.

> Is there another way to derive that an app is unmaintained from the metadata

I don't think so, only other way that I could imagine is checking if the repo is archived when it is on GH.

> Is there any way to find out if something has external dependencies (even if these get magically installed by default) in some of the appstore metadata

No, unfortunately not :/ I think the majority of apps work without any external depedency that isn't some kind of online service that is not self-hostable.

> Also, what's the growth rate of Nextcloud the appstore btw? If we're talking about <3 new apps per update from us we'd have to check for external dependencies, I might be OK with that

I think it's more than that, but also fluctuates a lot. One way to have an estimate is to look at https://github.com/nextcloud/app-certificate-requests since every new app needs such a certificate.

Given that we will only be able to automatically add apps that use the SPDX identifiers, we can still manually check all new ones since that migration will also be slow and probably a lot of apps will never migrate.
Unfortunately I don't think we can automate the testing for this a lot, since every app is different. Installing and enabling an app doesn't guarantee that it will actually work in the end.

**Comment by Ma27** [comment]
>  Unfortunately I don't think we can automate the testing for this a lot, since every app is different. Installing and enabling an app doesn't guarantee that it will actually work in the end.

Yeah, I don't think that's realistic either (and isn't nowadays already).
But this is essentially what I'd like to get more opinions on (though it boils down to whether we want to keep things as-is[1] or whether to package more and more stuff).

So for me the core question is whether we want to check apps for external dependencies that keep us from packaging[2].

[1] even though the testing story isn't that great already, I'd argue.
[2] and now that I'm thinking of it, we'll probably have to work in batches: there are 600 apps out there (assuming the number of merged PRs in the repo corresponds to the number of apps)

**Comment by provokateurin** [comment]
> there are 600 apps out there

Like I said we are limited to apps that use SPDX license identifiers, so it will be almost none at the moment and I would assume that the overall adoption will be very slow.

But given this, if we still need manual testing anyway to allow an app to be available, then the effort of adding SPDX licenses is not really relevant, since one could have already contacted the maintainers to clarify the license.

**Comment by Ma27** [comment]
> Like I said we are limited to apps that use SPDX license identifiers, so it will be almost none at the moment and I would assume that the overall adoption will be very slow.

I missed that part before, apologies!
Nonetheless I'd make sure to not add >20 apps per "batch" (just in case the adoption is quicker than expected).

So, we'd effectively need some tooling that adds (in batches) apps to our `apps.json` that are
* not there already
* have an spdx license specifier
* don't require additional packaging (i.e. manual audit)

Now, if nobody speaks up, I'm fine with that.
For me this has a rather low priority though given that I'm packed with a few other things. Would you be interested in doing that? Happy to help, of course.

**Comment by provokateurin** [comment]
Sounds good. I'd be happy to start the effort on this to get the tooling in place and then begin with the first apps.

**Comment by provokateurin** [comment]
I had some thoughts about this and I think it should be possible to automate the detection process for Nextcloud apps to use the SPDX identifiers. All official Nextcloud apps are adoption the REUSE specification, so it should be possible to automate checking if an app is compatible with -or-later or only -only.
We could use this to generate the files here, but I think it's better to do this upstream and change the license identifiers, to then build the NixOS tooling to accept the SPDX identifiers from info.xml.
